src/stroke.rs

```diff
@@ -0,0 +1,630 @@
+        self.start_cap = cap;
+        self.end_cap = cap;
+        self
+    }
+
+    /// Builder method for setting the dashing parameters.
+    pub fn with_dashes<P>(mut self, offset: f64, pattern: P) -> Self
+    where
+        P: IntoIterator,
+        P::Item: Borrow<f64>,
+    {
+        self.dash_offset = offset;
+        self.dash_pattern.clear();
+        self.dash_pattern
+            .extend(pattern.into_iter().map(|dash| *dash.borrow()));
+        self
+    }
+
+    /// Builder method for setting whether or not the stroke should be affected
+    /// by the scale of any applied transform.
+    pub fn with_scale(mut self, yes: bool) -> Self {
+        self.scale = yes;
+        self
+    }
+}
+
+/// Collection of values representing lengths in a dash pattern.
+pub type Dashes = SmallVec<[f64; 4]>;
+
+/// Internal structure used for creating strokes.
+#[derive(Default)]
+struct StrokeCtx {
+    // Probably don't need both output and forward, can just concat
+    output: BezPath,
+    forward_path: BezPath,
+    backward_path: BezPath,
+    start_pt: Point,
+    start_norm: Vec2,
+    start_tan: Vec2,
+    last_pt: Point,
+    last_tan: Vec2,
+    // if hypot < (hypot + dot) * bend_thresh, omit join altogether
+    join_thresh: f64,
+}
+
+/// Expand a stroke into a fill.
+pub fn stroke(path: impl IntoIterator<Item = PathEl>, style: &Stroke, tolerance: f64) -> BezPath {
+    if style.dash_pattern.is_empty() {
+        stroke_undashed(path, style, tolerance)
+    } else {
+        let dashed = DashIterator::new(path.into_iter(), style.dash_offset, &style.dash_pattern);
+        stroke_undashed(dashed, style, tolerance)
+    }
+}
+
+/// Version of stroke expansion for styles with no dashes.
+fn stroke_undashed(
+    path: impl IntoIterator<Item = PathEl>,
+    style: &Stroke,
+    tolerance: f64,
+) -> BezPath {
+    let mut ctx = StrokeCtx {
+        join_thresh: 2.0 * tolerance / style.width,
+        ..Default::default()
+    };
+    for el in path {
+        let p0 = ctx.last_pt;
+        match el {
+            PathEl::MoveTo(p) => {
+                ctx.finish(style);
+                ctx.start_pt = p;
+                ctx.last_pt = p;
+            }
+            PathEl::LineTo(p1) => {
+                if p1 != p0 {
+                    let tangent = p1 - p0;
+                    ctx.do_join(style, tangent);
+                    ctx.last_tan = tangent;
+                    ctx.do_line(style, tangent, p1);
+                }
+            }
+            PathEl::QuadTo(p1, p2) => {
+                if p1 != p0 && p2 != p0 {
+                    let q = QuadBez::new(p0, p1, p2);
+                    let (tan0, tan1) = PathSeg::Quad(q).tangents();
+                    ctx.do_join(style, tan0);
+                    ctx.last_tan = tan1;
+                    ctx.do_cubic(style, q.raise(), tolerance);
+                }
+            }
+            PathEl::CurveTo(p1, p2, p3) => {
+                if p1 != p0 && p2 != p0 && p3 != p0 {
```

Comment:
Per f2f this appears to be too strict. For example, there are valid cases where `p1 == p0`. Same above for the quad case.